go/signedexchange/signedexchange.go

```diff
@@ -0,0 +1,187 @@
+package signedexchange
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+
+	"github.com/WICG/webpackage/go/signedexchange/cbor"
+	"github.com/WICG/webpackage/go/signedexchange/mice"
+)
+
+type Input struct {
+	// Request
+	requestUri *url.URL
+
+	// Response
+	responseStatus int
+	responseHeader http.Header
+
+	// Payload
+	payload []byte
+}
+
+func NewInput(uri *url.URL, status int, headers http.Header, payload []byte, miRecordSize int) (*Input, error) {
+	i := &Input{
+		requestUri:     uri,
+		responseStatus: status,
+		responseHeader: headers,
+	}
+	if err := i.miEncode(payload, miRecordSize); err != nil {
+		return nil, err
+	}
+	return i, nil
+}
+
+func (i *Input) miEncode(payload []byte, recordSize int) error {
+	var buf bytes.Buffer
+	mi, err := mice.Encode(&buf, payload, recordSize)
+	if err != nil {
+		return err
+	}
+	i.payload = buf.Bytes()
+	i.responseHeader.Add("Content-Encoding", "mi-sha256")
+	i.responseHeader.Add("MI", mi)
+	return nil
+}
+
+// AddSignedHeadersHeader adds 'signed-headers' header to the response.
+//
+// Signed-Headers is a Structured Header as defined by
+// [I-D.ietf-httpbis-header-structure]. Its value MUST be a list (Section 4.8
+// of [I-D.ietf-httpbis-header-structure]) of lowercase strings (Section 4.2 of
+// [I-D.ietf-httpbis-header-structure]) naming HTTP response header fields.
+// Pseudo-header field names (Section 8.1.2.1 of [RFC7540]) MUST NOT appear in
+// this list.
+func (i *Input) AddSignedHeadersHeader(ks ...string) {
```

Comment:
Would it make sense for this tool to assume that all headers are signed? We support signing a subset to allow the signatures to survive intermediate proxies, and this tool doesn't need to be a proxy. Then you could remove this function, and have AddSignatureHeader also add the Signed-Headers header.

Reply:
I'm preparing a separate PR to address this. Will follow up shortly.

Reply:
https://github.com/hajimehoshi/webpackage/pull/3

Reply:
Done.